metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/execution/WorkflowExecutor.java

```diff
@@ -81,41 +78,31 @@ public WorkflowExecution call() {
     }
 
     //Cancel workflow and all other than finished plugins if the workflow was cancelled during execution
-    if (workflowExecutionDao.isCancelling(workflowExecution.getId())) {
+    //Check there is no finished date to make sure the workflow hasn't actually already finished
+    if (workflowExecutionDao.isCancelling(workflowExecution.getId()) && finishDate == null) {
       workflowExecution.setAllRunningAndInqueuePluginsToCancelled();
       LOGGER.info(
           "Cancelled running user workflow execution with id: {}", workflowExecution.getId());
     } else if (finishDate != null) { //finishedDate can be null in case of Exception
       workflowExecution.setFinishedDate(finishDate);
       workflowExecution.setWorkflowStatus(WorkflowStatus.FINISHED);
+      workflowExecution.setCancelling(false);
       LOGGER.info("Finished user workflow execution with id: {}", workflowExecution.getId());
     } else {
-      workflowExecution.checkAndSetAllRunningAndInqueuePluginsToFailedIfOnePluginHasFailed();
+      workflowExecution.checkAndSetAllRunningAndInqueuePluginsToCancelledIfOnePluginHasFailed();
     }
     //The only full update is used here. The rest of the execution uses partial updates to avoid losing the cancelling state field
     workflowExecutionDao.update(workflowExecution);
     return workflowExecution;
   }
 
-  private void runInQueueStateWorkflowExecution(RLock lock) {
-    workflowExecution.setStartedDate(new Date());
-    workflowExecution.setWorkflowStatus(WorkflowStatus.RUNNING);
-    workflowExecutionDao.updateMonitorInformation(workflowExecution);
-    lock.unlock(); //Unlock as soon as possible
-    AbstractMetisPlugin previousMetisPlugin = null;
-    for (AbstractMetisPlugin metisPlugin : workflowExecution.getMetisPlugins()) {
-      finishDate = runMetisPlugin(previousMetisPlugin, metisPlugin);
-      if (workflowExecutionDao.isCancelling(workflowExecution.getId())
-          || metisPlugin.getPluginStatus() == PluginStatus.FAILED) {
-        break;
-      }
-      previousMetisPlugin = metisPlugin;
-    }
-  }
-
-  private void runRunningStateWorkflowExecution(RLock lock) {
+  private void runInqueueOrRunningStateWorkflowExecution(RLock lock) {
     //Run if the workflowExecution was retrieved from the queue in RUNNING state. Another process released it and came back into the queue
-    workflowExecution.setUpdatedDate(new Date());
+    if (workflowExecution.getWorkflowStatus() == WorkflowStatus.RUNNING) {
+      workflowExecution.setUpdatedDate(new Date());
+    } else {
+      workflowExecution.setStartedDate(new Date());
+    }
     workflowExecution.setWorkflowStatus(WorkflowStatus.RUNNING);
     workflowExecutionDao.updateMonitorInformation(workflowExecution);
     lock.unlock(); //Unlock as soon as possible
```

Comment:
This seems like something that might go wrong in the future. It only works because nothing else happens after it before this is unlocked anyway in the calling method. May I suggest splitting this method in a locked part and an unlocked part? Then this statement can go and everything becomes a bit cleaner.